### ISSUE TITLE
At least one byte needs to be present for successful C / b decodes

### ISSUE THREAD

**Issue by kiropowered** [phpseclib/phpseclib#1970]
Hello,
I am actually using SFTP class to log into AWS Tranfer Family in a long running process that need to upload files.
After a while AWS seems like to disconnect the user and phpseclib does not detect it.

In order to check if the connection is still open I call isConnected method but it always return true and then after I try to upload a file it leads to the following error; 
Connection closed prematurely

I tried a different approach using ping()  and now the error is : 
At least one byte needs to be present for successful C / b decodes

This is the logs:
`-> NET_SSH2_MSG_REQUEST_FAILURE (since last: 0.0001, network: 0s)

<- NET_SSH2_MSG_DEBUG (since last: 0, network: 0s)
00000000  00:00:00:00:4e:53:46:54:50:3a:20:6b:65:79:20:6f  ....NSFTP: key o
00000010  70:74:69:6f:6e:73:3a:20:61:67:65:6e:74:2d:66:6f  ptions: agent-fo
00000020  72:77:61:72:64:69:6e:67:20:70:6f:72:74:2d:66:6f  rwarding port-fo
00000030  72:77:61:72:64:69:6e:67:20:70:74:79:20:75:73:65  rwarding pty use
00000040  72:2d:72:63:20:78:31:31:2d:66:6f:72:77:61:72:64  r-rc x11-forward
00000050  69:6e:67:00:00:00:00                             ing....

<- NET_SSH2_MSG_DEBUG (since last: 0, network: 0s)
00000000  00:00:00:00:4e:53:46:54:50:3a:20:6b:65:79:20:6f  ....NSFTP: key o
00000010  70:74:69:6f:6e:73:3a:20:61:67:65:6e:74:2d:66:6f  ptions: agent-fo
00000020  72:77:61:72:64:69:6e:67:20:70:6f:72:74:2d:66:6f  rwarding port-fo
00000030  72:77:61:72:64:69:6e:67:20:70:74:79:20:75:73:65  rwarding pty use
00000040  72:2d:72:63:20:78:31:31:2d:66:6f:72:77:61:72:64  r-rc x11-forward
00000050  69:6e:67:00:00:00:00                             ing....

<- NET_SSH2_MSG_UNIMPLEMENTED (since last: 0.004, network: 0.004s)
00000000  00:00:00:06                                      ....

At least one byte needs to be present for successful C / b decodes
0 /var/www/html/vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php(2664): phpseclib3\Common\Functions\Strings::unpackSSH2('C', false)
1 /var/www/html/vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php(2334): phpseclib3\Net\SSH2->privatekey_login('xxxxx', Object(phpseclib3\Crypt\RSA\PrivateKey))
2 /var/www/html/vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php(2279): phpseclib3\Net\SSH2->login_helper('xxxxx', Object(phpseclib3\Crypt\RSA\PrivateKey))
3 /var/www/html/vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php(2191): phpseclib3\Net\SSH2->sublogin('xxxxx', Object(phpseclib3\Crypt\RSA\PrivateKey))
4 /var/www/html/vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php(3386): phpseclib3\Net\SSH2->login('xxxxx', Object(phpseclib3\Crypt\RSA\PrivateKey))
5 /var/www/html/vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php(3368): phpseclib3\Net\SSH2->reconnect()`

---------------

I am using the version 3.0.33, tried also 3.0.34, but got the same errors

Other logs I got :
SSH_MSG_GLOBAL_REQUEST: hostkeys-00@openssh.com
SSH_MSG_DEBUG: SFTP: key options: agent-forwarding port-forwarding pty user-rc x11-forwarding
SSH_MSG_DEBUG: SFTP: key options: agent-forwarding port-forwarding pty user-rc x11-forwarding
SSH_MSG_GLOBAL_REQUEST: hostkeys-00@openssh.com
SSH_MSG_DEBUG: SFTP: key options: agent-forwarding port-forwarding pty user-rc x11-forwarding
SSH_MSG_DEBUG: SFTP: key options: agent-forwarding port-forwarding pty user-rc x11-forwarding
SSH_MSG_DISCONNECT: NET_SSH2_DISCONNECT_PROTOCOL_ERROR
data packet referred to nonexistent channel 0

**Comment by terrafrost** [comment]
Can you try the latest git version of 3.0?

Off hand I'm thinking that https://github.com/phpseclib/phpseclib/commit/404f86fb61fed2fc174ab7844252d2750c268f6e might fix the issue for you.

**Comment by kiropowered** [comment]
I tried with the version you mentionned but unfortunatly I still got the issue.

**Comment by terrafrost** [comment]
Well I'm not able to reproduce the issue.

```php
$key = PublicKeyLoader::load('...');

$ssh = new \phpseclib3\Net\SFTP('blahblah.server.transfer.us-west-2.amazonaws.com');
$ssh->login('phpseclib', $key);

echo $ssh->pwd() . "\n";

$ssh->disconnect();
$ssh->ping();

echo $ssh->pwd() . "\n";
```
That works just fine with the latest git version.

With 3.0.34 I got this error:

```
Fatal error: Uncaught phpseclib3\Exception\ConnectionClosedException: Connection closed by server
```

I tried to do `sleep(120);` instead of `$ssh->disconnect();`, hoping that I could get AWS to close the connection but it didn't.

Normally the way I test `ping` is by adding this to my /etc/ssh/sshd_config file:

```
ClientAliveInterval 1
ClientAliveCountMax 1
```
But clearly I can't do that with AWS.

So I'm curious...  does the above script that I shared, wherein I do `$ssh->disconnect();` and then `$ssh->ping();` right after reproduce the issue for you?

As for `isConnected`...  not really sure what I can do about that. `isConnected()` does `is_resource($this->fsock) && !feof($this->fsock)`. idk how else I'm supposed to know if `fwrite()` is going to fall without actually writing data, which is the whole point of `ping`. I mean I guess I could add `pingWithoutReconnect()` but idk. As for why you'd want to call `isConnected()` instead of `ping` or a hypothetical `pingWithoutReconnect` but, to be fair, `ping()` was added after the fact. It was inspired by https://www.php.net/mysqli-ping . `isConnected()` is the older method. Turning `isConnected()` ino `pingWithoutReconnect` feels like a BC break.

**Comment by kiropowered** [comment]
Hello, firstable I would like to thank you trying to figure out the issue in such a specific case.

I tried the code you provided

With branch 3.0, didn't get errors

With 3.0.34 I got this error as well:
`Fatal error: Uncaught phpseclib3\Exception\ConnectionClosedException: Connection closed by server`
Also a notice was triggered : 
`Notice: fputs(): Send of 36 bytes failed with errno=32 Broken pipe` on 
`/var/www/html/vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php","line":3809,"function":"send_binary_packet","class":"phpseclib3\\Net\\SSH2"`

-------
Then in a context of a long running process using sleep I was able to reproduce it on branch 3.0 with the following scenario

`
echo $ssh->pwd() . "\n";
$ssh->ping();
sleep(520);
$ssh->ping();
sleep(550);
$ssh->ping();
echo $ssh->pwd() . "\n";
`

It gives the error:
At least one byte needs to be present for successful C / b decodes

I tried with different sleep values but it's so random but this scenario fail each time.
I looked at AWS transfer family quotas and it says that the idle connection timeout is 1800, but it's clearly less and kinda random I think

**Comment by terrafrost** [comment]
I think https://github.com/phpseclib/phpseclib/commit/23f117e32b09ff3c9c1c049f991604014de7c68a should fix the issue.

Also, I think this should more reliably reproduce the issue:

```
$key = PublicKeyLoader::load('...');

$ssh = new \phpseclib3\Net\SFTP('blahblah.server.transfer.us-west-2.amazonaws.com');
$ssh->login('phpseclib', $key);

echo $ssh->pwd() . "\n";

$ssh->disconnect();
$ssh->ping();

echo $ssh->pwd() . "\n";

$ssh->disconnect();
$ssh->ping();

echo $ssh->pwd();
```

**Comment by kiropowered** [comment]
Awesome, well I didn't think about using disconnect as client because I was trying to reproduce my use case, but indeed it's much more reliable.
I initially thought it was specifically related to AWS because I didn't see any related issues in a long running process but I guess it's just due to consecutive disconnect and ping calls no matter sftp service.
Do you have any idea when do you plan to release this fix ?

**Comment by terrafrost** [comment]
I guess I can try to do one in a week or so!

**Comment by terrafrost** [comment]
I'm gonna try to get a fix in for #1972 before I do a new release btw.

Thank you for your patience!

**Comment by terrafrost** [comment]
3.0.35 has been released and contains this code change.

Thanks!
